align/constGap.go

```diff
@@ -3,62 +3,223 @@ package align
+				if j%checkersize == 0 {
+					trace_prep_j[j/checkersize][i] = mRowCurrent[j]
+				}
 			}
 		}
-		if i < mColumn-1 {
-			mRowPrevious, mRowCurrent = mRowCurrent, mRowPrevious //reuse mRow variables to save memory, swap slices rather than copy(mRowPrevious,mRowCurrent) for easy update, but only up until the second to last row
+
+		if i%checkersize == 0 && i < mColumn-1 {
+			copy(trace_prep_i[i/checkersize], mRowCurrent) //copy instead of assign values to avoid pointers automatically updating/synchronizing the 2 matrices
+			mRowPrevious, mRowCurrent = mRowCurrent, mRowPrevious
+		} else if i < mColumn-1 {
+			mRowPrevious, mRowCurrent = mRowCurrent, mRowPrevious //reuse mRow variables to save memory, but only up until the second to last row
 		}
 	}
 
-	//write cigar
-	route := make([]Cigar, 1)
-	for i, j, routeIdx = mColumn-1, len(mRowCurrent)-1, 0; i > 0 || j > 0; {
-		if route[routeIdx].RunLength == 0 {
-			route[routeIdx].RunLength = 1
-			route[routeIdx].Op = trace[i][j]
-		} else if route[routeIdx].Op == trace[i][j] {
-			route[routeIdx].RunLength += 1
+	score_highest := mRowCurrent[len(mRowCurrent)-1]
+	score_highest_i := mColumn - 1
+	score_highest_j := len(mRowCurrent) - 1
+	return score_highest, score_highest_i, score_highest_j, trace_prep_i, trace_prep_j
+}
+
+//Step 2
+func FillTraceback(alpha []dna.Base, beta []dna.Base, scores [][]int64, gapPen int64, checkersize int, score_highest_i int, score_highest_j int, trace_prep_i [][]int64, trace_prep_j [][]int64, k1 int, k2 int, i_inChecker_min_Previous int, j_inChecker_min_Previous int) ([][]ColType, int, int) {
+
+	mRowCurrent := make([]int64, len(beta)+1)
+	mRowPrevious := make([]int64, len(beta)+1)
+	mRowPrevious = trace_prep_i[k1]
+	//i_inChecker and j_inChecker are internal (within checkerboard) versions of i and j, e.g. for checkersize=3, i_inChecker can take on the values 0, 1, 2
+	//i_max and j_max are external (non-inChecker) versions of i_inChecker_max and j_inChecker_max
+	var i, j, i_inChecker, j_inChecker, i_inChecker_max, j_inChecker_max, i_max, j_max int
+	if i_inChecker_min_Previous >= 0 {
+		//use the i_inChecker_min_Previous from the previous k1/k2 loop iteration's Step 3 WriteCigar to find the i_max at which to stop filling the trace matrix
+		i_max = checkersize*k1 + 1 + i_inChecker_min_Previous
+	} else {
+		//otherwise, since no restriction from previous k1/k2 loop Step 3 WriteCigar, find the minimum value between "where the next checkerboard starts" and "the i of the highest score", and that is where to stop filing the trace matrix
+		i_max = int(math.Min(float64(checkersize*(k1+1)), float64(score_highest_i)))
+	}
+	if j_inChecker_min_Previous >= 0 {
+		j_max = checkersize*k2 + 1 + j_inChecker_min_Previous
+	} else {
+		j_max = int(math.Min(float64(checkersize*(k2+1)), float64(score_highest_j)))
+	}
+	i_inChecker_max = (i_max - 1) % checkersize
+	j_inChecker_max = (j_max - 1) % checkersize
+	trace := make([][]ColType, checkersize)
```

Comment:
One thing to think about in the future would be passing in this trace matrix and then passing it back again so that it can be re-used each time.

Reply:
Changed step 2 inputs to pass trace back and forth to reuse memory